In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
#creating path to import the files
path = r'C:\Users\rbrand\New York Citi Bikes'

In [8]:
#Importing citibike.csv data set from original data.
cb_original_data = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'citibike.csv'), index_col=False)

In [10]:
#Reviwing dataframe consistency: 
cb_original_data.shape

(50000, 18)

In [12]:
cb_original_data.head()

trip_id  bike_id weekday  start_hour       start_time  start_station_id  \
0  LnQzQk    16013     Mon          18   9/9/2013 18:18               523   
1  IL9boN    15230     Thu          18  9/12/2013 18:38               257   
2  46clGB    17942     Wed          19  9/18/2013 19:44               479   
3  v7vdFt    19683     Sat          11  9/28/2013 11:54               527   
4  VGBsb5    18024     Sat          18   9/7/2013 18:08               521   

        start_station_name  start_station_latitude  start_station_longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

          end_time  end_station_id     end_station_name  end_station_latitude  \
0   9/9/2013 18:35             334      W 20 St & 7 Ave             40.742388   
1  9/12/2013 18:48             236  St Marks Pl & 2 Ave             40.728419   
2  9/18/2013 19:50             513     W 56 St & 10 Ave             40.768254   
3  9/28/2013 12:03             441      E 52 St & 2 Ave             40.756014   
4   9/7/2013 18:46             476      E 31 St & 3 Ave             40.743943   

   end_station_longitude  trip_duration      subscriber  birth_year  gender  
0             -73.997262            993      Subscriber      1968.0       2  
1             -73.987140            581      Subscriber      1983.0       1  
2             -73.988639            361      Subscriber      1989.0       1  
3             -73.967416            561      Subscriber      1988.0       2  
4             -73.979661           2296  Non-Subscriber         NaN       0

In [13]:
# Reviewing missing values
missing_values = cb_original_data.isnull().sum()
print(missing_values)

trip_id                       0
bike_id                       0
weekday                       0
start_hour                    0
start_time                    0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_time                      0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
trip_duration                 0
subscriber                    0
birth_year                 6979
gender                        0
dtype: int64


In [14]:
#Calculating the median of birth_year
median_birth_year = cb_original_data['birth_year'].median()

In [16]:
# Print the median
print("Median birth year:", median_birth_year)

Median birth year: 1978.0


In [20]:
# Adding missing values with the median.
cb_original_data['birth_year'].fillna(median_birth_year, inplace=True)

In [21]:
#Double checking missing values
missing_values = cb_original_data.isnull().sum()
print(missing_values)

trip_id                    0
bike_id                    0
weekday                    0
start_hour                 0
start_time                 0
start_station_id           0
start_station_name         0
start_station_latitude     0
start_station_longitude    0
end_time                   0
end_station_id             0
end_station_name           0
end_station_latitude       0
end_station_longitude      0
trip_duration              0
subscriber                 0
birth_year                 0
gender                     0
dtype: int64


In [24]:
#checking if there is any duplicates based on trip_id 
trip_id_duplicates = cb_original_data[cb_original_data.duplicated(subset=['trip_id'], keep=False)]

# Print the duplicate rows based on trip_id
print("Duplicate rows based on trip_id:")
print(trip_id_duplicates)

Duplicate rows based on trip_id:
Empty DataFrame
Columns: [trip_id, bike_id, weekday, start_hour, start_time, start_station_id, start_station_name, start_station_latitude, start_station_longitude, end_time, end_station_id, end_station_name, end_station_latitude, end_station_longitude, trip_duration, subscriber, birth_year, gender]
Index: []


It seems that there are no duplicate rows based on the trip_id column in the DataFrame. This indicates that each trip_id is unique in the dataset, which is a good sign for data integrity.

In [26]:
# Findign the most common value in the gender column
most_common_gender = cb_original_data['gender'].mode()[0]
print(most_common_gender)

1


In [27]:
# Replacing '0' values with the most common value (1)
cb_original_data['gender'].replace(0, most_common_gender, inplace=True)

In [28]:
# Confirm that '0' values have been replaced
count_0_values_after = (cb_original_data['gender'] == 0).sum()
print("Number of '0' values after replacement:", count_0_values_after)

Number of '0' values after replacement: 0


In [31]:
# Handling Incorrect Trip Durations
# Converting trip duration from seconds to minutes for better interpretability
cb_original_data['trip_duration_minutes'] = cb_original_data['trip_duration'] / 60

In [33]:
# Remove outliers: Assuming realistic trips are between 1 minute and 24 hours (1440 minutes)
cb_original_data = cb_original_data[(cb_original_data['trip_duration_minutes'] >= 1) & (cb_original_data['trip_duration_minutes'] <= 1440)]

In [34]:
# Checking summary statistics after cleaning
summary_statistics = cb_original_data['trip_duration_minutes'].describe()
print(summary_statistics)

count    50000.000000
mean        13.983048
std          9.561067
min          1.000000
25%          6.950000
50%         11.200000
75%         18.533333
max         44.950000
Name: trip_duration_minutes, dtype: float64


It seems like the summary statistics for the trip_duration_minutes column after cleaning are as follows:

- Count: 50,000 (indicating the number of non-null values after cleaning)
- Mean: Approximately 13.98 minutes
- Standard Deviation: Approximately 9.56 minutes
- Minimum: 1 minute
- 25th Percentile (Q1): Approximately 6.95 minutes
- Median (50th Percentile or Q2): Approximately 11.20 minutes
- 75th Percentile (Q3): Approximately 18.53 minutes
- Maximum: 44.95 minutes

These statistics provide valuable insights into the distribution of trip durations after cleaning. It appears that most trips have durations ranging from approximately 1 minute to 18.53 minutes, with some longer trips up to 44.95 minutes. This information can defintely guide me further analysis and decision-making regarding trip durations in your dataset.

In [36]:
#Exporting cb_clean_data. 
cb_original_data.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'cb_clean_data.csv'))